.. _contents:

#####################################
Welcome to PyCogent3's documentation!
#####################################

**Contents**

.. toctree::
    :maxdepth: 1
    
    install
    README
    coding_guidelines
    data_file_links
    examples/index
    cookbook/index
    developer_notes
    licenses
    ChangeLog


.. todolist::

********************
WARNING & DISCLAIMER
********************

Be warned that PyCogent3_ is a significantly changed library from the original `PyCogent <http://www.pycogent.org>`_. Most of these changes involve elimination of modules, conversion to `PEP8 <https://www.python.org/dev/peps/pep-0008/>`_ compliance and other refactors. These changes are still underway!

Please see the `wiki pages <https://bitbucket.org/pycogent3/cogent3/wiki/Home>`_ for a detailed changelog on the API changes.

*************
YOU CAN HELP!
*************

Posting Bugs
============

If you discover a bug, especially something that worked in PyCogent but not in PyCogent3_, please `post a ticket <https://bitbucket.org/pycogent3/cogent3/issues>`_!

When posting a ticket, please provide a minimum working example to reproduce the issue. Also include the versions of the library and other tools (e.g. attach the result of ``$ pip freeze > libs.txt``).

Getting Involved
================

We would greatly appreciate assistance in updating the project to PEP8, or anything else you think needs doing.

Please post a ticket, or comment on an existing one, indicating your intention so we can assist. Then it's the usual "fork", "pull request" dance.

********
Overview
********

PyCogent3 is a software library for genomic biology. It is a fully integrated and thoroughly tested framework for: conducting novel probabilistic analyses of biological sequence evolution; and generating publication quality graphics. It is distinguished by many unique built-in capabilities (such as true codon alignment) and the frequent addition of entirely new methods for the analysis of genomic data.

Our primary goal is to provide a collection of rigourously validated tools for the manipulation and analysis of genome biology data sets.

.. todo::

    update the literature cited

*******
Support
*******

We acknowledge the provision by `Wingware <http://wingware.com>`_ of free licenses for their professional IDE. Their committment, over more than a decade, to supporting our development of Open Source software for science is greatly appreciated. (The port of PyCogent to Python 3 was made much easier by using Wing's refactor tools!)

********
Citation
********

If you use this software for published work please cite -- `Knight et al., 2007, Genome Biol, 8, R171 <http://genomebiology.com/2007/8/8/R171>`_.

******
Search
******

* :ref:`search`

.. _PyCogent3: https://bitbucket.org/pycogent3/cogent3

*******
Support
*******

We acknowledge the provision by `Wingware <http://wingware.com>`_ of free licenses for their professional IDE. Their committment, over more than a decade, to supporting our development of Open Source software for science is greatly appreciated. (The port of PyCogent to Python 3 was made much easier by using Wing's refactor tools!)

**********************
News and Announcements
**********************

PyCogent3 News and Announcements are available via the News and Announcements Blog at http://pycogent.wordpress.com.

`Subscribe to the News and Annoucements RSS Feed <http://pycogent.wordpress.com/feed>`_
